### PR TITLE
Route hotkeys to all visible controls

### PR DESCRIPTION
**Pull request by mstv**
## Proposed changes

- Route hotkeys to the other visible controls in FormBrowse and FormCommit if not handled by the focused control 

## Test methodology 

- manual

## Test environment(s) 

- Git Extensions 33.33.33
- Build 9916393dc46aaeb1c5a5006d046768b301bd5721
- Git 2.31.1.windows.1 (recommended: 2.33.0 or later)
- Microsoft Windows NT 10.0.19044.0
- .NET 5.0.12
- DPI 96dpi (no scaling)

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/CommandsDialogs/FormBrowse.cs [hunk 1]
```diff
@@ -2134,6 +2134,16 @@ private void QuickFetch()
...
         }
 
+        protected override bool ProcessCmdKey(ref Message msg, Keys keyData)
+        {
+            return !IsDesignMode && HotkeysEnabled
+                && (base.ProcessCmdKey(ref msg, keyData) // upstream
+                    || (!GitExtensionsControl.IsTextEditKey(keyData) // downstream (without keys for quick search)
+                        && (RevisionGridControl.ProcessHotkey(keyData)
+                            || (CommitInfoTabControl.SelectedTab == DiffTabPage && revisionDiff.ProcessHotkey(keyData))
+                            || (CommitInfoTabControl.SelectedTab == TreeTabPage && fileTree.ProcessHotkey(keyData)))));
```
Comment (gerhardol) [inline GitUI/CommandsDialogs/FormBrowse.cs:2144]: Scripts?

File: GitUI/CommandsDialogs/FormBrowse.cs [hunk 2]
```diff
@@ -2134,6 +2134,16 @@ private void QuickFetch()
             UICommands.RepoChangedNotifier.Notify();
         }
 
+        protected override bool ProcessCmdKey(ref Message msg, Keys keyData)
+        {
+            return !IsDesignMode && HotkeysEnabled
+                && (base.ProcessCmdKey(ref msg, keyData) // upstream
```
Comment (gerhardol) [inline GitUI/CommandsDialogs/FormBrowse.cs:2140]: Handle downstrem first?
Generic

File: ResourceManager/GitExtensionsControl.cs [hunk 3]
```diff
@@ -74,6 +74,18 @@ public virtual void TranslateItems(ITranslation translation)
 
         #region Hotkeys
 
+        /// <summary>Checks if a hotkey wants to handle the key and execute it.</summary>
```
Comment (gerhardol) [inline ResourceManager/GitExtensionsControl.cs:77]: Can you describe the relation to `ProcessCmdKey` a little here?
I get the overview but cannot suggest something...

**Review by gerhardol — COMMENTED** [review]
+1
LGTM, have not run (or considered all caveeats)

File: GitUI/CommandsDialogs/FormBrowse.cs (hunk 2 above)
Comment (mstv) [reply GitUI/CommandsDialogs/FormBrowse.cs:2140]: The sequence is:
- current control
- parents (form)
- new: other visible controls of a form which supports it

Perhaps, I should explain "upstream" / "downstream" in the comment.
Would "upstream (parent form)" suffice?

Comment (gerhardol) [reply GitUI/CommandsDialogs/FormBrowse.cs:2140]: parents form (upstream) is better I believe

File: GitUI/CommandsDialogs/FormBrowse.cs (hunk 1 above)
Comment (mstv) [reply GitUI/CommandsDialogs/FormBrowse.cs:2144]: The hotkeys of scripts are added to the `HotKeys` property of all controls and forms by `HotkeySettingsManager.LoadHotkeys`.

File: ResourceManager/GitExtensionsControl.cs (hunk 3 above)
Comment (mstv) [reply ResourceManager/GitExtensionsControl.cs]: done

File: ResourceManager/GitExtensionsControl.cs [hunk 4]
```diff
@@ -74,7 +74,12 @@ public virtual void TranslateItems(ITranslation translation)
 
         #region Hotkeys
 
-        /// <summary>Checks if a hotkey wants to handle the key and execute it.</summary>
+        /// <summary>
+        /// Checks if a hotkey wants to handle the key and execute that hotkey
+        /// (without propagating an unhandled key to the base class function <cref>ProcessCmdKey</cref>).
+        /// Can be overridden in order to execute a hotkey of other (visible) subcontrols
+        /// if this focused/queried control does not have such an (active) hotkey itself.
```
Comment (gerhardol) [inline ResourceManager/GitExtensionsControl.cs:81]: I do not understand the last line.

**Review by gerhardol — COMMENTED** [review]
+1
Still have not run

File: ResourceManager/GitExtensionsControl.cs (hunk 4 above)
Comment (mstv) [reply ResourceManager/GitExtensionsControl.cs:81]: > I do not understand the last line.

This function `ProcessHotkey` is called on key press by the function `Control.ProcessCmdKey` (if the control is *focused*)
or when one of its parent controls asks / *queries* whether this control has an (active) hotkey.
(If a hotkey is not active, the key is treated as "not handled yet" and is propagated to the other controls and upstream.)

How would you reword the comment?

Comment (gerhardol) [reply ResourceManager/GitExtensionsControl.cs:81]:
```suggestion
        /// Checks if the control wants to handle the key and execute that hotkey
        /// (without propagating an unhandled key to the base class function as in <cref>ProcessCmdKey</cref>).
        /// Can be overridden in order to execute a hotkey for a (visible) subcontrol
        /// (if this focused/queried control does not have such an (active) hotkey itself).
```

File: GitUI/CommandsDialogs/RevisionDiffControl.cs [hunk 5]
```diff
@@ -1194,6 +1194,15 @@ public void SwitchFocus(bool alreadyContainedFocus)
             }
         }
 
+        public override bool ProcessHotkey(Keys keyData)
+        {
+            return base.ProcessHotkey(keyData) // generic handling of this controls's hotkeys (upstream)
+                || (!GitExtensionsControl.IsTextEditKey(keyData) // downstream (without keys for quick search and filter)
+                    && ((DiffFiles.Visible && DiffFiles.ProcessHotkey(keyData))
+                        || (DiffText.Visible && DiffText.ProcessHotkey(keyData))
+                        || (BlameControl.Visible && BlameControl.ProcessHotkey(keyData))));
```
Comment (gerhardol) [inline GitUI/CommandsDialogs/RevisionDiffControl.cs:1203]: Similar is needed for FileTree subcontrols

File: ResourceManager/GitExtensionsControl.cs (hunk 4 above)
Comment (mstv) [reply ResourceManager/GitExtensionsControl.cs]: Thank you, done.

File: GitUI/CommandsDialogs/RevisionDiffControl.cs (hunk 5 above)
Comment (mstv) [reply GitUI/CommandsDialogs/RevisionDiffControl.cs:1203]: Added for RevisionFileTreeControl.

**Review by gerhardol — APPROVED** [review]

**Comment by mstv** [comment]
Would like to merge this in three days or so.

**Comment by gerhardol** [comment]
With this PR all built in shortcuts must be handled as hotkeys.
For instance will Ctrl-A in the FileViewer select all files in the first group.
If you assign Ctrl-C to a hotkey then copy will not work either...

**Pull request merged**